### PR TITLE
feat(test-runner): implement `expect.poll`

### LINKED ISSUE #10235: [Feature] introduce predicate web-first assertions
Context: https://github.com/microsoft/playwright/issues/10187

In general, this is a feature request to support predicate as an argument to `expect`:

```js
await expect(asyncPredicate).toBe(10);
```

The `await expect(..)` construct should retry predicate until it reaches certain value.

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch implements `expect.poll()` method that polls given
predicate until a given synchronous matcher succeeds.

Usage:

```js
// wait until page gets 3 frames.
await expect.poll(() => page.frames().length, {
  timeout: 1000,
  message: 'custom error message',
}).toBe(3);
```

Fixes #10235

### CODE REVIEW

**Comment by aslushnikov** [comment]
Prior art & discussion: https://github.com/microsoft/playwright/pull/12614

File: packages/playwright-test/src/expect.ts [hunk 1]
```diff
@@ -240,7 +287,19 @@ function wrap(matcherName: string, matcher: any) {
...
 
     try {
-      const result = matcher.call(this, ...args);
+      let result;
+      const [receivedOrGenerator, ...otherArgs] = args;
+      if (isPoll) {
+        if (typeof receivedOrGenerator !== 'function')
+          throw new Error('`expect.poll()` accepts only function as a first argument');
+        if ((customMatchers as any)[matcherName] || matcherName === 'resolves' || matcherName === 'rejects')
+          throw new Error(`\`expect.poll()\` does not support ${matcherName} matcher.`);
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:296]:
```suggestion
          throw new Error(`\`expect.poll()\` does not support "${matcherName}"`);
```

File: packages/playwright-test/src/expect.ts [hunk 2]
```diff
@@ -240,7 +287,19 @@ function wrap(matcherName: string, matcher: any) {
...
+        result = pollMatcher(matcher, currentExpectTimeout({ timeout: pollTimeout }), this, receivedOrGenerator, ...otherArgs);
+      } else {
+        if (typeof receivedOrGenerator === 'function')
+          throw new Error('Cannot accept function as a first argument; use `expect.poll()` instead');
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:300]:
```suggestion
          throw new Error('Cannot accept function as a first argument; did you mean to use expect.poll()?');
```

File: packages/playwright-test/src/expect.ts [hunk 3]
```diff
@@ -180,6 +195,36 @@ class ExpectMetaInfoProxyHandler {
   }
 }
 
+function pollMatcher(matcher: any, timeout: number, thisArg: any, generator: () => any, ...args: any[]) {
+  return new Promise(async resolve => {
+    let result: { pass: boolean, message: () => string } | undefined = undefined;
+    const startTime = monotonicTime();
+    const pollIntervals = [0, 100, 250, 500];
+    while (monotonicTime() - startTime <= timeout) {
+      const received = await generator();
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:204]: Should we race this one?

```js
try {
  await expect.poll(() => page.locator('non-existing').textContent(), { timeout: 3000 }).toBe('whatever');
} catch (e) {
  // jk, go on!
}
```

File: packages/playwright-test/types/testExpect.d.ts [hunk 4]
```diff
@@ -28,8 +28,9 @@ type MakeMatchers<T, ReturnValue = T> = PlaywrightTest.Matchers<ReturnValue> &
   ExtraMatchers<T, APIResponse, APIResponseMatchers>
 
 export declare type Expect = {
-  <T = unknown>(actual: T, message?: string): MakeMatchers<T>;
-  soft: <T = unknown>(actual: T, message?: string) => MakeMatchers<T>;
+  <T = unknown>(actual: T, messageOrOptions?: string | { message?: string }): MakeMatchers<T>;
+  soft: <T = unknown>(actual: T, messageOrOptions?: string | { message?: string }) => MakeMatchers<T>;
+  poll: <T = unknown>(actual: () => T, messageOrOptions?: string | { message?: string, timeout?: number }) => Omit<PlaywrightTest.Matchers<T>, 'rejects' | 'resolves'>;
```
Comment (dgozman) [inline packages/playwright-test/types/testExpect.d.ts:33]:
```suggestion
  poll: <T = unknown>(actual: () => T | Promise<T>, messageOrOptions?: string | { message?: string, timeout?: number }) => Omit<PlaywrightTest.Matchers<T>, 'rejects' | 'resolves'>;
```

**Review by dgozman — APPROVED** [review]

File: packages/playwright-test/src/expect.ts [hunk 5]
```diff
@@ -240,7 +287,19 @@ function wrap(matcherName: string, matcher: any) {
     };
 
     try {
-      const result = matcher.call(this, ...args);
+      let result;
+      const [receivedOrGenerator, ...otherArgs] = args;
+      if (isPoll) {
+        if (typeof receivedOrGenerator !== 'function')
+          throw new Error('`expect.poll()` accepts only function as a first argument');
+        if ((customMatchers as any)[matcherName] || matcherName === 'resolves' || matcherName === 'rejects')
```
Comment (mxschmitt) [inline packages/playwright-test/src/expect.ts:295]: lets add a test for that?

[the remark this answers is not in the document]
Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts:298]: it does; added a test.

Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts:298]: there's a test for custom matchers!

File: packages/playwright-test/src/expect.ts [hunk 6]
```diff
@@ -180,6 +196,38 @@ class ExpectMetaInfoProxyHandler {
   }
 }
 
+function pollMatcher(matcher: any, timeout: number, thisArg: any, generator: () => any, ...args: any[]) {
+  return new Promise(async resolve => {
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:200]: nit: why not just an async function here?

**Review by dgozman — APPROVED** [review]

File: packages/playwright-test/src/expect.ts (hunk 6 above)
Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts]: fair. fixed!

File: packages/playwright-test/src/expect.ts [hunk 7]
```diff
@@ -89,21 +91,23 @@ export const printReceivedStringContainExpectedResult = (
...
-function createExpect(actual: unknown, message: string|undefined, isSoft: boolean) {
-  if (message !== undefined && typeof message !== 'string')
-    throw new Error('expect(actual, optionalErrorMessage): optional error message must be a string.');
-  return new Proxy(expectLibrary(actual), new ExpectMetaInfoProxyHandler(message || '', isSoft));
+function createExpect(actual: unknown, messageOrOptions: string|undefined | { message?: string, timeout?: number}, isSoft: boolean, isPoll: boolean) {
+  return new Proxy(expectLibrary(actual), new ExpectMetaInfoProxyHandler(messageOrOptions, isSoft, isPoll));
 }
 
 export const expect: Expect = new Proxy(expectLibrary as any, {
   apply: function(target: any, thisArg: any, argumentsList: [actual: unknown, message: string|undefined]) {
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts]: Fair - done.

File: packages/playwright-test/src/expect.ts [hunk 8]
```diff
@@ -133,19 +137,29 @@ const customMatchers = {
 };
 
 type ExpectMetaInfo = {
```
[the remark this answers is not in the document]
Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts:141]: This I did not follow!

File: packages/playwright-test/src/expect.ts [hunk 9]
```diff
@@ -133,19 +139,29 @@ const customMatchers = {
...
   isSoft: boolean;
+  isPoll: boolean;
+  pollTimeout?: number;
 };
 
 let expectCallMetaInfo: undefined|ExpectMetaInfo = undefined;
 
 class ExpectMetaInfoProxyHandler {
-  private _message: string;
+  private _message?: string;
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:151]: Do a single `this._info`.

File: packages/playwright-test/src/expect.ts [hunk 10]
```diff
@@ -180,6 +198,33 @@ class ExpectMetaInfoProxyHandler {
   }
 }
 
+async function pollMatcher(matcher: any, timeout: number, thisArg: any, generator: () => any, ...args: any[]) {
+  let result: { pass: boolean, message: () => string } | undefined = undefined;
+  const startTime = monotonicTime();
+  const pollIntervals = [100, 250, 500];
+  while (monotonicTime() - startTime <= timeout) {
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:205]: `timeout` could be 0.

File: tests/page/page-screenshot.spec.ts [hunk 11]
```diff
@@ -726,7 +726,8 @@ it.describe('page screenshot animations', () => {
     ]);
   });
 
-  it('should respect fonts option', async ({ page, server, isWindows }) => {
+  it('should respect fonts option', async ({ page, server, isWindows, browserName, isLinux }) => {
+    it.fixme(browserName === 'webkit' && isLinux, 'https://github.com/microsoft/playwright/issues/12839');
```
Comment (dgozman) [inline tests/page/page-screenshot.spec.ts:730]: Need a rebase?

**Review by dgozman — APPROVED** [review]

**Pull request merged**